File: day-22a/src/main.py
from collections.abc import Iterator, Sequence
from dataclasses import dataclass
from itertools import product, starmap
from operator import attrgetter
from typing import NamedTuple, Self, final


class XYPoint(NamedTuple):
    x: int
    y: int


@final
@dataclass(kw_only=True)
class Brick:
    min_x: int
    max_x: int
    min_y: int
    max_y: int
    min_z: int
    max_z: int

    def __post_init__(self) -> None:
        self._x_range = range(self.min_x, self.max_x + 1)
        self._y_range = range(self.min_y, self.max_y + 1)

    def z_range(self) -> range:
        return range(self.min_z, self.max_z + 1)

    def __repr__(self) -> str:
        return (
            f"Brick("
            f"x=({self.min_x}->{self.max_x}), "
            f"y=({self.min_y}->{self.max_y}), "
            f"z=({self.min_z}->{self.max_z})"
            f")"
        )

    def fall_by_one(self) -> None:
        self.min_z -= 1
        self.max_z -= 1

    def xy_points(self) -> Iterator[XYPoint]:
        return starmap(XYPoint, product(self._x_range, self._y_range))

    @classmethod
    def from_string(cls, string: str) -> Self:
        left, _, right = string.partition("~")
        x0, y0, z0 = map(int, left.split(","))
        x1, y1, z1 = map(int, right.split(","))
        return cls(
            min_x=min(x0, x1),
            max_x=max(x0, x1),
            min_y=min(y0, y1),
            max_y=max(y0, y1),
            min_z=min(z0, z1),
            max_z=max(z0, z1),
        )

    def __hash__(self) -> int:
        return id(self)


type ZCoordinate = int
type GridOfGrids = dict[ZCoordinate, dict[XYPoint, Brick]]


@dataclass(slots=True, frozen=True, kw_only=True)
class PuzzleInput:
    bricks: Sequence[Brick]
    map: GridOfGrids

    @classmethod
    def load(cls, *, input_filename: str) -> Self:
        with open(input_filename) as f:
            bricks = tuple(Brick.from_string(line) for line in f)
        grid_of_grids: GridOfGrids = {}
        for brick in bricks:
            for z in brick.z_range():
                grid = grid_of_grids.setdefault(z, {})
                grid.update(dict.fromkeys(brick.xy_points(), brick))
        return cls(bricks=bricks, map=grid_of_grids)


def drop_brick(brick: Brick, *, map: GridOfGrids) -> None:
    while brick.min_z > 1:
        grid_below = map.setdefault(brick.min_z - 1, {})
        if grid_below and any(point in grid_below for point in brick.xy_points()):
            break
        highest_grid = map[brick.max_z]
        for point in brick.xy_points():
            grid_below[point] = brick
            del highest_grid[point]
        brick.fall_by_one()


def has_two_or_more_bricks_below(
    brick: Brick, *, map: GridOfGrids
) -> bool:
    grid_below = map[brick.min_z - 1]
    bricks_below: set[Brick] = set()
    for point in brick.xy_points():
        brick_below: Brick | None = grid_below.get(point)
        if brick_below is not None:
            bricks_below.add(brick_below)
        if len(bricks_below) > 1:
            return True
    return False


def brick_could_safely_be_disintegrated(brick: Brick, *, map: GridOfGrids) -> bool:
    grid_above = map.setdefault(brick.max_z + 1, {})
    if not grid_above:
        return True
    bricks_above: set[Brick] = set()
    for point in brick.xy_points():
        brick_above: Brick | None = grid_above.get(point)
        if brick_above is not None:
            bricks_above.add(brick_above)
    return all(
        has_two_or_more_bricks_below(brick_above, map=map)
        for brick_above in bricks_above
    )


def solve(input_filename: str) -> int:
    puzzle_input = PuzzleInput.load(input_filename=input_filename)
    for brick in sorted(puzzle_input.bricks, key=attrgetter("min_z")):
        drop_brick(brick, map=puzzle_input.map)
    return sum(
        1
        for brick in puzzle_input.bricks
        if brick_could_safely_be_disintegrated(brick, map=puzzle_input.map)
    )


def main() -> None:
    print(solve("input.txt"))


if __name__ == "__main__":
    main()

File: day-22a/src/main.rs
use std::cell::RefCell;
use std::cmp::{max, min};
use std::collections::{HashMap, HashSet};
use std::fs::read_to_string;
use std::ops::RangeInclusive;
use std::{fmt::Display, str::FromStr};

use anyhow::{bail, Result};
use itertools::{iproduct, Itertools};

#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
struct XYPoint {
    x: u16,
    y: u16,
}

#[derive(Debug, Clone, Copy)]
struct Brick {
    min_x: u16,
    max_x: u16,
    min_y: u16,
    max_y: u16,
    min_z: u16,
    z_range_len: u16,
}

impl Brick {
    fn max_z(&self) -> u16 {
        self.min_z + self.z_range_len - 1
    }

    fn z_range(&self) -> RangeInclusive<u16> {
        self.min_z..=self.max_z()
    }

    fn fall_by_one(self) -> Self {
        Self {
            min_z: self.min_z - 1,
            ..self
        }
    }

    fn xy_points(&self) -> impl Iterator<Item = XYPoint> {
        let Self {
            min_x,
            max_x,
            min_y,
            max_y,
            ..
        } = *self;
        iproduct!(min_x..=max_x, min_y..=max_y).map(|(x, y)| XYPoint { x, y })
    }
}

impl Display for Brick {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Self {
            min_x,
            max_x,
            min_y,
            max_y,
            min_z,
            ..
        } = *self;
        let max_z = self.max_z();
        write!(
            f,
            "Brick(x=({min_x}->{max_x}), y=({min_y}->{max_y}), z=({min_z}->{max_z})"
        )
    }
}

impl FromStr for Brick {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        let [left, right] = s.split('~').collect_vec()[..] else {
            bail!("Expected the string to contain a `~`!")
        };
        let (x0, y0, z0) = match left.split(',').collect_vec()[..] {
            [x0, y0, z0] => (x0.parse()?, y0.parse()?, z0.parse()?),
            _ => bail!("Expected the left-hand side to contain exactly two commas!"),
        };
        let (x1, y1, z1) = match right.split(',').collect_vec()[..] {
            [x1, y1, z1] => (x1.parse()?, y1.parse()?, z1.parse()?),
            _ => bail!("Expected the right-hand side to contain exactly two commas!"),
        };
        let (min_z, max_z) = (min(z0, z1), max(z0, z1));
        Ok(Self {
            min_x: min(x0, x1),
            max_x: max(x0, x1),
            min_y: min(y0, y1),
            max_y: max(y0, y1),
            min_z,
            z_range_len: (max_z - min_z) + 1,
        })
    }
}

type ZCoordinate = u16;
type BrickId = u16;
type GridOfGrids = RefCell<HashMap<ZCoordinate, HashMap<XYPoint, BrickId>>>;
type IdToBrickMap = HashMap<BrickId, Brick>;

struct PuzzleInput {
    id_to_brick_map: IdToBrickMap,
    coord_to_brick_map: GridOfGrids,
}

impl FromStr for PuzzleInput {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        let mut id_to_brick_map: IdToBrickMap = HashMap::new();
        for (i, line) in s.lines().enumerate() {
            let i = i.try_into()?;
            let brick = line.parse()?;
            id_to_brick_map.insert(i, brick);
        }
        let coord_to_brick_map: GridOfGrids = RefCell::new(HashMap::new());
        for (id, brick) in &id_to_brick_map {
            for z in brick.z_range() {
                coord_to_brick_map.borrow_mut().entry(z).or_default();
                for point in brick.xy_points() {
                    coord_to_brick_map
                        .borrow_mut()
                        .get_mut(&z)
                        .unwrap()
                        .insert(point, *id);
                }
            }
        }
        Ok(Self {
            id_to_brick_map,
            coord_to_brick_map,
        })
    }
}

fn parse_input(input_filename: &str) -> Result<PuzzleInput> {
    let input = read_to_string(input_filename)?;
    PuzzleInput::from_str(&input)
}

fn drop_brick(brick_id: u16, mut brick: Brick, map: &mut GridOfGrids) -> Brick {
    while brick.min_z > 1 {
        let next_z_down = &(brick.min_z - 1);
        map.borrow_mut().entry(*next_z_down).or_default();
        if !map.borrow().get(next_z_down).unwrap().is_empty()
            && brick
                .xy_points()
                .any(|p| map.borrow().get(next_z_down).unwrap().contains_key(&p))
        {
            break;
        };
        for point in brick.xy_points() {
            map.borrow_mut()
                .get_mut(next_z_down)
                .unwrap()
                .insert(point, brick_id);
            map.borrow_mut()
                .get_mut(&(brick.max_z()))
                .unwrap()
                .remove(&point);
        }
        brick = brick.fall_by_one()
    }
    brick
}

fn has_two_or_more_bricks_below(
    brick: &Brick,
    map: &HashMap<ZCoordinate, HashMap<XYPoint, BrickId>>,
) -> bool {
    let mut brick_ids_below = HashSet::new();
    let grid_below = map.get(&(brick.min_z - 1)).unwrap();
    for point in brick.xy_points() {
        if let Some(brick_id_below) = grid_below.get(&point) {
            brick_ids_below.insert(brick_id_below);
        }
        if brick_ids_below.len() > 1 {
            return true;
        }
    }
    false
}

fn brick_could_safely_be_disintegrated(
    brick: &Brick,
    map: &GridOfGrids,
    id_to_brick_map: &IdToBrickMap,
) -> bool {
    let mut map_borrow = map.borrow_mut();
    let Some(grid_above) = map_borrow.get_mut(&(brick.max_z() + 1)) else {
        return true;
    };
    let mut brick_ids_above = HashSet::new();
    for point in brick.xy_points() {
        if let Some(brick_id_above) = grid_above.get(&point) {
            brick_ids_above.insert(brick_id_above);
        }
    }
    brick_ids_above
        .iter()
        .all(|id| has_two_or_more_bricks_below(id_to_brick_map.get(id).unwrap(), &map.borrow()))
}

fn solve(input_filename: &str) -> usize {
    let input = RefCell::new(parse_input(input_filename).unwrap());
    let mut bricks = {
        let borrowed_input = input.borrow();
        Vec::from_iter(
            borrowed_input
                .id_to_brick_map
                .iter()
                .map(|(id, brick)| (*id, *brick)),
        )
    };
    bricks.sort_unstable_by_key(|(_, brick)| brick.min_z);
    let bricks = bricks
        .iter()
        .map(|(id, brick)| drop_brick(*id, *brick, &mut input.borrow_mut().coord_to_brick_map));
    let mut answer = 0;
    for brick in bricks {
        let input = input.borrow();
        if brick_could_safely_be_disintegrated(
            &brick,
            &input.coord_to_brick_map,
            &input.id_to_brick_map,
        ) {
            answer += 1;
        }
    }
    answer
}

fn main() {
    println!("{}", solve("input.txt"))
}
